### ISSUE TITLE
we should make semantically meaningless positional arguments positional only in our operator API

### ISSUE THREAD

**Issue by dagitses** [pytorch/pytorch#99351]
### 🚀 The feature, motivation and pitch

See #99265 for an example of some of the chaos this causes. Position-only arguments were added in Python 3.8 and we recently dropped support for 3.7 so now is the first opportunity to do this.

Prior art: the Python array API is aggressively using position only arguments.

Removing this will be BC breaking, so here is a rough idea for how it should be done.
1. add support for doing this in our codegen
2. add a linter to prevent new uses of things like tensor1, tensor2, etc.
3. add a warning that fires when calling existing functions with named arguments
4. remove warning and migrate to position only arguments after a release

@mruberry @rgommers

**Comment by rgommers** [comment]
Yes, positional-only arguments would be great to have, +1 from me.

**Comment by leofang** [comment]
+1

**Comment by albanD** [comment]
Adding this for new functions sounds pretty cool!
I'm not sure it is realistic to do the full BC-breaking change though. We would need to look into details at usage to make sure this is not going to lead to widespread breakage.

**Comment by rgommers** [comment]
My $2c on that is that it should be checked indeed, but widespread breakage seems unlikely, and a little bit of breakage (after deprecation warnings, as in @dagitses' plan) seems like a price worth paying for higher code quality and less untested code paths. You can write something like this today:
```python
>>> t = torch.arange(5)
>>> torch.add(input=t, other=t)
tensor([0, 2, 4, 6, 8])
```
but that doesn't make much sense and I bet many of these pass-as-keyword usages don't have any test coverage.

As a data point: in numpy and scipy we've made parameters positional-only by accident a number of times, because the numpy ufunc machinery enforces that. I can't remember a single bug report about it.

**Comment by gchanan** [comment]
I'd also add checking for usages of the keyword args on github as a step.  If there are more than a handful for a particular case, we can not do the change.  If there are only a handful, we can send PRs to update.

**Comment by asmeurer** [comment]
It might be useful to outline a set of guidelines for when arguments should be positional-only and when arguments should allowed to be passed as a keyword. I'm also thinking here for NumPy 2.0 where will will be making a similar change.

For instance, an argument that specifies an tensor/array should pretty much always be positional-only, but things like `dims`/`axis` should be passable as keyword (and etc. There's a lot of argument types throughout the API). Additionally, some arguments should be keyword-only, e.g., `out` appears to be always keyword-only in `torch` APIs.

**Comment by dagitses** [comment]
I agree having these guidelines would be valuable. And it would be best to standardize the guidelines to the extent that is practical across the different API implementations.

I do think there are exceptions where we want tensors specified as keywords, e.g. tensors that are used as masks.

I would say, non-controversially I presume, for operators that exist with the pair `x.op(*args, **kwargs)` and `op(x, *args, **kwargs)`, `x` should not be passable as a kwarg.

Some simple general principles:
 * bool args should almost always be keywords
 * any arguments that we would have named x1, x2, etc. should never be keywords :)
 * single argument operators should probably never take a keyword argument, and the argument's semantics should almost always be obvious from the name of the function.

Most of these guidelines are probably not specific to PyTorch or numpy, and are likely to be good practice in Python in general.
